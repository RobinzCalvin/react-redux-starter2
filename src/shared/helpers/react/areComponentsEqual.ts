function areComponentsEqual(a: React.ComponentType, b: React.ComponentType): boolean {
  return (a.displayName || a.name) === (b.displayName || b.name);
}

export default areComponentsEqual;
