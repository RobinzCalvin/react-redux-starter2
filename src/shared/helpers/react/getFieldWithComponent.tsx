import React from 'react';
import { Field, FieldRenderProps, FieldProps as RFFieldProps } from 'react-final-form';
import { O } from 'ts-toolbelt';

type BaseWrappedFieldProps = FieldRenderProps & {
  value?: any;
  onChange?: any;
};

// keys for pick props from RFFieldProps, Omit dont work because RFFieldProps have index signature
type RFFieldPropKey =
  | 'allowNull' | 'format' | 'formatOnBlur' | 'parse' | 'name'
  | 'isEqual' | 'subscription' | 'validate' | 'value';

function getFieldWithComponent<P extends BaseWrappedFieldProps>(
  Component: React.ComponentType<P>,
  type?: string,
) {
  type OwnProps = Omit<P, keyof BaseWrappedFieldProps>;
  type FieldProps = Pick<RFFieldProps, RFFieldPropKey>;
  type ResultProps = O.Merge<FieldProps, OwnProps>;

  const result: React.FC<ResultProps> = (props: ResultProps) => (
    <Field type={type} {...props} component={Component} />
  );

  result.displayName = `FieldWithComponent(${Component.displayName || Component.name || 'Component'})`;
  return result;
}

export { getFieldWithComponent };
