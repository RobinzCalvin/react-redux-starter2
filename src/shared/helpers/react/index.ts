export { default as withProps } from './withProps';
export { default as getFieldWithComponent } from './getFieldWithComponent';
export { default as withComponent } from './withComponent';
export { default as areComponentsEqual } from './areComponentsEqual';
