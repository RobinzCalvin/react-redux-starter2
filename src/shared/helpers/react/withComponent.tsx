import React from 'react';
import { Omit } from 'react-redux';

interface IBaseProps {
  component?: React.ReactType<any>;
}

type StringType = keyof JSX.IntrinsicElements;

type ExtractProps<T extends StringType | React.ComponentType<any>> =
  T extends StringType
    ? JSX.IntrinsicElements[T]
    : T extends React.ComponentType<any>
      ? React.ComponentProps<T>
      : never;

function withComponent<C extends StringType | React.ComponentType<any>>(component: C) {
  return function decorate<BProps extends IBaseProps>(BaseComponent: React.ComponentType<BProps>) {
    type CProps = ExtractProps<C>;

    return function WithComponent(props: Omit<BProps, 'component'> & CProps) {
      return <BaseComponent component={component} {...props} />;
    };
  };
}

export { withComponent };
