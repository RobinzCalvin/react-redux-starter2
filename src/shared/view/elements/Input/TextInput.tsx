import React from 'react';
import { autobind } from 'core-decorators';

import TextField, { TextFieldProps } from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';
import IconButton from '@material-ui/core/IconButton';

import Adjust from '@material-ui/icons/Adjust';

type IProps = Omit<TextFieldProps, 'ref'>;

interface IState {
  type?: string;
}

class TextInput extends React.PureComponent<IProps, IState> {
  public state: IState = {
    type: this.props.type,
  };

  public render() {
    const { type } = this.state;

    return (
      <TextField
        {...this.props as TextFieldProps}
        type={type}
        InputProps={{
          ...this.props.InputProps,
          endAdornment: this.renderEndAdornment(),
        }}
        fullWidth
      />
    );
  }

  private renderEndAdornment(): React.ReactNode {
    const { type } = this.props;

    if (type === 'password') {
      return (
        <InputAdornment position="end">
          <IconButton
            aria-label="Toggle password visibility"
            onClick={this.handleClickShowPassword}
          >
            <Adjust />
          </IconButton>
        </InputAdornment>
      );
    }
  }

  @autobind
  private handleClickShowPassword() {
    this.setState(state => ({
      type: state.type === 'password' ? 'text' : 'password',
    }));
  }
}

export { IProps };
export default TextInput;
