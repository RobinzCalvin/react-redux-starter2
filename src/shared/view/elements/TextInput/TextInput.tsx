import * as React from 'react';
import { FormControl, FormControlProps } from 'react-bootstrap';
import FormEventHandler = React.FormEventHandler;
import Component = React.Component;
import FormEvent = React.FormEvent;

interface Props {
  type: string;
  placeholder?: string;
  name?: string;
  minLength?: number;
  maxLength?: number;
  min?: number;
  max?: number;
  step?: number;
  pattern?: string;
  hidden?: boolean;
  onChange?: FormEventHandler<Component<FormControlProps, {}>>;
}

type EventType = FormEvent<Component<ControlProps, {}>>;
type ControlProps = FormControlProps;

function TextInput (props: Props) {
  return (
    <FormControl
      type={props.type}
      placeholder={props.placeholder}
      name={props.name}
      minLength={props.minLength}
      maxLength={props.maxLength}
      step={props.step}
      pattern={props.pattern}
      onChange={props.onChange}
    />
  );
}

export { Props, ControlProps, EventType };
export default TextInput;
