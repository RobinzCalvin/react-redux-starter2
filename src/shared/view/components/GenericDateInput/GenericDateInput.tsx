import * as React from 'react';
import { bind } from 'decko';
import { IProps as GenericFieldProps } from '../GenericInput/GenericInput';
import TextInput, { EventType } from './../../elements/TextInput/TextInput';
import InputGroup from './../../elements/InputGroup/InputGroup';
import Errors from '../../elements/Errors/Errors';
import FormEvent = React.FormEvent;
import Component = React.Component;

type IProps = GenericFieldProps;
interface IState {
  errors: string[];
  isEdited: boolean;
}

class GenericDateInput extends React.PureComponent<IProps, IState> {
  private standardHTMLpattern = '^([0-9]{4})-([0-9]{2})-([0-9]{2})$';
  private errors = {
    invalid: 'Incorrect Date format',
    required: 'Field is required',
  };

  constructor(props: GenericFieldProps) {
    super(props);
    this.state = {
      errors: [],
      isEdited: false,
    };
  }

  public componentDidMount() {
    this.validateAndChange(''); // first changing with validation, but without showing errors
  }

  public render() {
    const {
      name,
      label,
      pattern,
      placeholder,
    } = this.props;
    const { errors, isEdited } = this.state;

    return (
      <InputGroup label={label}>
        <TextInput
          type="date"
          name={name}
          placeholder={placeholder}
          pattern={pattern}
          onChange={this.onChange}
        />
        <Errors errors={this.props.errors ? errors.concat(this.props.errors) : errors} hidden={!isEdited} />
      </InputGroup>
    );
  }

  @bind
  private onChange(event: EventType) {
    const value: string = (event.nativeEvent.target as HTMLInputElement).value;
    this.validateAndChange(value);
    this.setState((prevState: IState) => ({ ...prevState, isEdited: true }));
  }

  @bind
  private validateAndChange(rawValue: string): void {
    const { pattern, required, onChange } = this.props;
    const errors: string[] = [];
    let value: string;

    if ((new RegExp(this.standardHTMLpattern)).test(rawValue)) {
      // current browser support date inputs
      value = rawValue.split('-').reverse().join('-');
    } else {
      value = rawValue;
    }

    if (pattern && !(new RegExp(pattern)).test(value)) {
      errors.push(this.errors.invalid);
    }

    if (required && !value.length) {
      errors.push(this.errors.required);
    }

    this.setState({ ...this.state, errors });

    if (onChange) {
      onChange(value, errors);
    }
  }
}

export { IProps, IState };
export default GenericDateInput;
