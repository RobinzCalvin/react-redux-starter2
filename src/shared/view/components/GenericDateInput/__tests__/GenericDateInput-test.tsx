import * as React from 'react';
import { expect } from 'chai';
import { mount, shallow, ShallowWrapper } from 'enzyme';
import { spy } from 'sinon';
import GenericDateInput, { IProps, IState } from './../GenericDateInput';
import { IProps as ITextInputProps } from 'shared/view/elements/TextInput/TextInput';

describe('(Shared) View', () => {
  describe('(Component) GenericDateInput', () => {
    it('should render <input> tag', () => {
      const component = mount(<GenericDateInput />);
      const header = component.find('input');

      expect(header.length).to.be.equal(1);
    });

    it('should call onChange prop, if onChange called in TextInput composite component', () => {
      const onChangeSpy = spy();
      const component = shallow<IProps, IState>(<GenericDateInput onChange={onChangeSpy} />);
      const textInput = component.find('TextInput').first() as ShallowWrapper<ITextInputProps, {}>;
      // const textInputProps: ITextInputProps = textInput.props();
      const event = { nativeEvent: { target: { value: '123' } } } as any;
      textInput.simulate('change', event);

      expect(onChangeSpy.calledTwice).to.be.true;
      expect(onChangeSpy.firstCall.args).to.deep.equal(['', []]);
      expect(onChangeSpy.secondCall.args).to.deep.equal(['123', []]);
    });

    it('should call onChange, when just mounted with \'Field required\' error, if component is required', () => {
      const onChangeSpy = spy();
      mount<IProps, IState>(<GenericDateInput onChange={onChangeSpy} required />);
      expect(onChangeSpy.calledOnce).to.be.true;
      expect(onChangeSpy.firstCall.args).to.deep.equal(['', ['Field is required']]);
    });
  });
});
