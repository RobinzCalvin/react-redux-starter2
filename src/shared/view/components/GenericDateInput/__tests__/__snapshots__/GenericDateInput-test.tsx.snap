// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Shared) View (Component) GenericDateInput should match snapshot 1`] = `
"<InputGroup label={[undefined]}>
  <TextInput type=\\"date\\" name={[undefined]} placeholder={[undefined]} pattern={[undefined]} onChange={[Function]} />
  <Errors errors={{...}} hidden={true} />
</InputGroup>"
`;
